### PR TITLE
Fix meson defaults

### PR DESCRIPTION
**Pull request by haampie**
Use `debugoptimized` aka `-O2 -g` as a default, just like CMake's `RelWithDebInfo` default

Do not add `--strip` unconditionally, cause it strips debug symbols :(

Add a variant for `default_library`, which conveniently supports `both` too.

### CODE REVIEW

File: lib/spack/spack/build_systems/meson.py
Comment (haampie) [inline lib/spack/spack/build_systems/meson.py:146]:
```suggestion
        * ``-Dbuildtype``
        * ``-Dstrip``
        * ``-Ddefault_library``
```

**Review by alalazo — DISMISSED** [review]
LGTM. Waiting for a second green light before merging since it changes a couple of previous defaults.

**Comment by alalazo** [comment]
@tmerrick1

**Comment by adamjstewart** [comment]
Do the docs need to be updated too?

**Comment by haampie** [comment]
I've updated the docs a bit.

**Comment by haampie** [comment]
Another ping: @tmerrick1

**Comment by tmerrick1** [comment]
I guess that I am surprised that you want debug as the default mode.

**Comment by haampie** [comment]
> I guess that I am surprised that you want debug as the default mode.

Here's my considerations:

1. It's useful to have debug symbols `-g` to get reasonable stack traces / core dumps;
2. Performance is not sacrificed other than potentially going from `-O3` to `-O2` by default -- debugoptimized doesn't imply `-O0` / `-Og` but rather debug symbols.
3. Using `debugoptimized` would bring consistency with CMakePackage's `RelWithDebInfo` default -- there are 700+ cmake packages vs 17 meson packages, so changing meson to use release with debug info by default would be the smallest change.

Would you want to comment if you're in favor / opposed to this change? It's not explicit in your comment

**Comment by tmerrick1** [comment]
I am in favor of it.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
